sycl/include/CL/sycl/detail/queue_impl.hpp

```diff
@@ -57,6 +58,12 @@ class queue_impl {
   }
 
   ~queue_impl() {
```

Comment:
Hm... Mr. Stroustrup says that throwing exceptions from the destructor is a bad design.
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c37-make-destructors-noexcept

Should we expect `async_handler` to throw? If not, should we add `noexcept` specifier to `async_handler`/queue and context destructors? or avoid calling it from the destructor?
+@mkinsner, @jbrodman

Reply:
It seems throwing exception from async_handler is the only C++ way for the user to handle async exceptions. Otherwise user needs to use some "global" object and check the status after throw_asynchronous method is done.

Reply:
This is probably not the right place to discuss this topic, but C++11 provides "Thread support library" with the following facilities:

> **Futures**
> The standard library provides facilities to obtain values that are returned and to catch exceptions that are thrown by asynchronous tasks (i.e. functions launched in separate threads). These values are communicated in a shared state, in which the asynchronous task may write its return value or store an exception, and which may be examined, waited for, and otherwise manipulated by other threads that hold instances of std::future or std::shared_future that reference that shared state.

(from https://en.cppreference.com/w/cpp/thread)

AFAIK, standard library objects do not throw from their destructors.

Reply:
Having troubles imaging how futures help here.